### PR TITLE
Windowing Support for the DaskRunner

### PR DESCRIPTION
**Pull request by alxmrs**
This CL adds basic Windowing support to this runner, including a few tests for side inputs.

Take two of https://github.com/apache/beam/pull/23913. It looks like some of the issues in CI causing this dask change to fail were resolved on their own with changes in master. 

Reviewers: @jrmccluskey

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: .github/workflows/dask_runner_tests.yml [hunk 1]
```diff
@@ -65,7 +65,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
+        os: [ubuntu-latest, macos-latest]
```
Comment (alxmrs) [inline .github/workflows/dask_runner_tests.yml:68]: I could use help getting this to run on windows.

**Comment by alxmrs** [comment]
Hey Jack: With windowing support implemented, does that mean that we can run streaming pipelines "for free"?

**Comment by alxmrs** [comment]
Jack: I'm seeing a weird failure in a test that shouldn't ever happen. 

https://github.com/apache/beam/actions/runs/5634715157/job/15264903865?pr=27618#step:7:160

Any ideas what could be going on?

**Comment by jrmccluskey** [comment]
To answer the streaming question: IIRC you'll also need to support SDFs and checkpointing for streaming to work in a meaningful way. 

Not much experience with Windows dev at this point, I'm not the person to ask on that one.

**Comment by Abacn** [comment]
> Jack: I'm seeing a weird failure in a test that shouldn't ever happen.
> 
> https://github.com/apache/beam/actions/runs/5634715157/job/15264903865?pr=27618#step:7:160
> 
> Any ideas what could be going on?

This is irrelevant flaky test. It happened before I remember, due to a unit test modified the global RuntimeValueProvider and did not change it back in tearDown, causing subsequent unrelated test could fail. Rerun the tasks now all passed

**Comment by jrmccluskey** [comment]
Also going to loop in @lostluck, since they have very current runner-writing experience

**Comment by lostluck** [comment]
Honestly I don't think I'll be much help here either. "Wrapping" an existing runner is different from "making it work from scratch" like I've done with prism.

I don't know anything about dask so I can only describe what's needed in abstract.

If Dask (or this wrapper) is able to track and propagate watermarks down the pipeline, then windowing is a straightforward triggering of applicable aggregations.

Otherwise, for prism, i found it impossible to map Stream Execution to a simple batch execution loop since the watermarks weren't able to behave right. So prism has an element manager that tracks available data and watermarks for stages.

So, the question is then: does this dask wrapper track watermarks?

File: sdks/python/apache_beam/runners/dask/dask_runner.py [hunk 2]
```diff
@@ -159,6 +163,8 @@ def is_fnapi_compatible():
     return False
 
   def run_pipeline(self, pipeline, options):
+    import dask
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply sdks/python/apache_beam/runners/dask/dask_runner.py:166]: The try..except is a recent addition (to deal with a test-time-only threading issue). Nested imports are what I've had for a long time. That's the main reason for why things are the way they are. 

Nested imports are there so that users don't have to depend on dask to load the runner.

**Comment by alxmrs** [comment]
I may misunderstand what watermarks are, but I believe this PR does this, see this file: sdks/python/apache_beam/runners/dask/transform_evaluator.py

**Comment by lostluck** [comment]
To set expectations, I don't have the time to decipher python until after the 2.50 release goes out. So, September ish.

**Comment by jrmccluskey** [comment]
I can decipher the python, just wanted runner-aware eyes around as well so I don't miss something important

**Comment by jrmccluskey** [comment]
@alxmrs https://beam.apache.org/documentation/glossary/#watermark

Effectively, the internal representation of what time the runner thinks it has processed all the data up to. This adds concepts of late data along with an understanding of when a runner is "safe" to emit the data for a given window

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py [hunk 3]
```diff
@@ -65,15 +123,54 @@ def apply(self, input_bag: OpInput) -> db.Bag:
 class ParDo(DaskBagOp):
```
Comment (jrmccluskey) [inline sdks/python/apache_beam/runners/dask/transform_evaluator.py:123]: Docstrings for the classes and methods would be helpful

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py [hunk 4]
```diff
@@ -83,21 +180,20 @@ def key(item):
...
-  def apply(self, input_bag: OpInput) -> db.Bag:
+  def apply(self, input_bag: t.List[db.Bag]) -> db.Bag:
     assert type(input_bag) is list, 'Must take a sequence of bags!'
     return db.concat(input_bag)
 
 
 TRANSLATIONS = {
     _Create: Create,
     apache_beam.ParDo: ParDo,
-    apache_beam.Map: Map,
```
Comment (jrmccluskey) [inline sdks/python/apache_beam/runners/dask/transform_evaluator.py:100]: Could you explain why Map was removed in this PR? Doesn't seem to be hurting anything based on testing, just curious

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py [hunk 5]
```diff
@@ -25,15 +25,73 @@
...
+    assign_context = WindowFn.AssignContext(item.timestamp, item.value)
+    windowed_value = WindowedValue(
+        item.value, item.timestamp, tuple(window_fn.assign(assign_context)))
+  else:
+    windowed_value = WindowedValue(item, 0, (GlobalWindow(), ))
+
+  return windowed_value
+
+
+def defenestrate(x):
```
Comment (jrmccluskey) [inline sdks/python/apache_beam/runners/dask/transform_evaluator.py:68]: This might be the best function name I've ever seen

Comment (cisaacstern) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:68]: incredible. also what if `x` is passed as `'prague'` ... do we get some kind of easter egg behavior?

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py (hunk 4 above)
Comment (alxmrs) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:100]: I found via experiments that this was never exercised! `Map` translates to a composite of `ParDo` every time.

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py [hunk 6]
```diff
@@ -83,21 +180,20 @@ def key(item):
...
-      return k, [elm[1] for elm in v]
+      return k, [defenestrate(elm[1]) for elm in v]
 
     return input_bag.groupby(key).map(value)
 
 
 class Flatten(DaskBagOp):
-  def apply(self, input_bag: OpInput) -> db.Bag:
+  def apply(self, input_bag: t.List[db.Bag]) -> db.Bag:
     assert type(input_bag) is list, 'Must take a sequence of bags!'
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:190]: Good to know!

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py (hunk 5 above)
Comment (alxmrs) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:68]: :)

**Comment by jrmccluskey** [comment]
Run PythonFormatter PreCommit

**Comment by jrmccluskey** [comment]
Run PythonLint PreCommit

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alxmrs) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:214]: Nit: this should be `isinstance()`

[the remark this answers is not in the document]
Comment (cisaacstern) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py:214]: > Nit: this should be isinstance()

Good call @alxmrs, I can fix that in my next push.

File: sdks/python/apache_beam/runners/dask/dask_runner.py [hunk 7]
```diff
@@ -177,6 +183,6 @@ def run_pipeline(self, pipeline, options):
 
     dask_visitor = self.to_dask_bag_visitor()
     pipeline.visit(dask_visitor)
-
-    futures = client.compute(list(dask_visitor.bags.values()))
+    opt_graph = dask.optimize(list(dask_visitor.bags.values()))
```
Comment (cisaacstern) [inline sdks/python/apache_beam/runners/dask/dask_runner.py:186]: In https://github.com/apache/beam/pull/27618/commits/305699bbbdde9b555ecd88c4bedc0b312a268a53, I've cast `dask_visitor.bags.values()` to `list` (as is done in master), because I found that the subsequent call to `client.compute` would silently fail otherwise.

The **Details** below contain the closest thing I have to an MRE of why casting to `list()` is necessary here. I still don't understand is why the tests in `dask_runner_test.py` still appear to pass without casting to `list()` 🤔 .  

<details>

```pycon
In [1]: import dask.bag; import distributed

In [2]: b = dask.bag.from_sequence([1, 2, 3]).map(lambda x: x*2).map(print)

In [3]: opt_b = dask.optimize(b)

In [4]: with distributed.Client() as c:
   ...:     futures = c.compute(opt_b)
   ...:     distributed.wait(futures)
   ...: 
2
4
6

In [5]: bdict = {"bag": b}

In [6]: opt_bdict_values = dask.optimize(bdict.values())

In [7]: with distributed.Client() as c:
   ...:     futures = c.compute(opt_bdict_values)
   ...:     distributed.wait(futures)
   ...: 

In [8]: opt_bdict_values_aslist = dask.optimize(list(bdict.values()))

In [9]: with distributed.Client() as c:
   ...:     futures = c.compute(opt_bdict_values_aslist)
   ...:     distributed.wait(futures)
   ...: 
2
4
6
```
</details>

File: sdks/python/apache_beam/runners/dask/transform_evaluator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cisaacstern) [reply sdks/python/apache_beam/runners/dask/transform_evaluator.py]: Fixed in https://github.com/apache/beam/pull/27618/commits/32ee6e13b1d1853b5760ef480d76da581ddadb79

**Comment by cisaacstern** [comment]
Status update: @alxmrs invited me to wrap this PR up for him, which I am delighted to do as this is will be quite useful for us over in [Pangeo Forge](https://pangeo-forge.readthedocs.io). So far I have resolved the pre-commit failures and started to familiarize myself with the implementation. AFAICT, the major remaining work is getting codecov to pass through addition of further tests. I will continue on that and ping the reviewers here when I'm ready for re-review.

Thanks all in advance for your patience with me as a first-time Beam contributor. I've been an enthusiastic user of the Python SDK for a year plus now, and am glad to have an opportunity to contribute! 🙏

File: sdks/python/apache_beam/runners/dask/dask_runner.py (hunk 7 above)
Comment (cisaacstern) [reply sdks/python/apache_beam/runners/dask/dask_runner.py]: Reverted this in https://github.com/apache/beam/pull/27618/commits/9557e3aba4b67ca85d770bcfd9cc677affce2178 ... I wish I could explain why I did in the first place but looks like my test env was just deeply messed up in some unknown way 🤦 . Ok, onward!

**Comment by cisaacstern** [comment]
In working on testing, I stumbled across the somewhat alarming fact that something in this PR is resulting in `assert_that` _**always passing**_, including when asked to evaluate plainly falsey expressions such as `0 == 1`, e.g. the following `test_zero_equals_one` should obviously fail, as it does on the `DirectRunner` parametrization, but running with the `DaskRunner` (from this PR) results in an `XPASS` (i.e. unexpected pass):

```python
import apache_beam as beam
import pytest
from apache_beam.runners.dask.dask_runner import DaskRunner
from apache_beam.testing import test_pipeline
from apache_beam.testing.util import assert_that, equal_to

runners = ["DirectRunner", DaskRunner()]
runner_ids = ["DirectRunner", "DaskRunner"]


@pytest.mark.xfail(strict=True)  # These tests **SHOULD FAIL** because 0 != 1
@pytest.mark.parametrize("runner", runners, ids=runner_ids)
def test_zero_equals_one(runner):
   with test_pipeline.TestPipeline(runner=runner) as p:
       pcoll = p | beam.Create([0])
       assert_that(pcoll, equal_to([1]))
```

I wrote up https://github.com/cisaacstern/dask-windowing-assert-that-broken/pull/1 as a gut check that this was a regression unique to this PR, and the CI there convinces me that this issue does not effect the DaskRunner as released in `2.51.0`:

Screen Shot 2023-10-19 at 12 43 57 PM

Certainly open to any suggestions from others as to what may be causing this, but I don't yet feel blocked here, just will need a bit more time to dig into what is causing this, and then will report back here with my findings.

**Comment by cisaacstern** [comment]
Ok, narrowing in on the issue here! Looks like this _does_ relate to the "to cast to list or not to cast to list" discussion I had with myself starting in https://github.com/apache/beam/pull/27618#discussion_r1357577926. As shown here, _not_ casting to list results in no computation actually happening:

```python
# bag.py

import dask.bag as db
import distributed as ddist

bag = db.from_sequence([1])
bag_dict = {"bag": bag}

if __name__ == "__main__":
    with ddist.Client() as c:
        bag_result = c.gather(c.compute(bag))
        print(f"{bag_result=}")

        dict_vals_result = c.gather(c.compute(bag_dict.values()))
        print(f"{dict_vals_result=}")

        dict_vals_to_unpacked_list_result = c.gather(c.compute(*list(bag_dict.values())))
        print(f"{dict_vals_to_unpacked_list_result=}")
```
```console
➜ python bag.py
bag_result=[1]
dict_vals_result=dict_values([dask.bag<from_sequence, npartitions=1>])
dict_vals_to_unpacked_list_result=[1]
```
So looks like casting to list is definitely the way to go. I think I backed off that change because it caused a serialization issue (which caused tests to fail). So I'll re-implement the list casting now and re-examine what that serialization issue was, and report back.

**Comment by cisaacstern** [comment]
Getting closer. To confirm, the following change (which I haven't yet pushed) is definitely required to get computable graphs:

```diff
--- a/sdks/python/apache_beam/runners/dask/dask_runner.py
+++ b/sdks/python/apache_beam/runners/dask/dask_runner.py
@@ -183,6 +183,6 @@ class DaskRunner(BundleBasedDirectRunner):
 
     dask_visitor = self.to_dask_bag_visitor()
     pipeline.visit(dask_visitor)
-    opt_graph = dask.optimize(dask_visitor.bags.values())
+    opt_graph = dask.optimize(*list(dask_visitor.bags.values()))
     futures = client.compute(opt_graph)
     return DaskRunnerResult(client, futures)
```

With this change, very basic pipelines do run end-to-end without error, but any pipeline that involves a pipeline stage that includes a _**deeply-nested function or object definition**_ raises some form of:

```python-traceback
  File ".../distributed/distributed/client.py", line 3159, in _graph_to_futures
    header, frames = serialize(ToPickle(dsk), on_error="raise")
  File ".../distributed/distributed/protocol/serialize.py", line 378, in serialize
    raise TypeError(msg, str_x) from exc
TypeError: ('Could not serialize object of type HighLevelGraph', '<ToPickle: HighLevelGraph with 2 layers.\n<dask.highlevelgraph.HighLevelGraph object at 0x159b6d990>\n 0. 5800487232\n 1. bf0ad4e590a57329d7d581b1a8a01c79\n>')
```

This is with the latest `distributed==2023.10.0`. Deeply nested function/object definitions are common throughout the Python SDK, and significantly, appear in the definition of `assert_that`, which appears to be why `assert_that` is failing.

According to the dask docs [here](https://distributed.dask.org/en/stable/serialization.html#defaults):

> Computational tasks like f(x) that are defined and serialized on client processes and deserialized and run on worker processes. These are serialized using a fixed scheme decided on by those libraries. Today this is a combination of pickle and cloudpickle.

So it seems like if we can get this serialization onto the cloudpickle serialization code path, it might work? (Since IIUC cloudpickle handles deeply nested objects better than pickle.) AFAICT, I am not hitting the cloudpickle code path here, but some further digging is necessary:

- Can I make an MRE which demonstrates this pickling issue using vanilla Dask (no Beam)?
- If is, it is resolved with cloudpickle?
- And finally, is the current attempt to serialize the HighLevelGraph indeed missing the cloudpickle path, and if so how can we can we get it onto that path?

**Comment by cisaacstern** [comment]
It seems we are hitting the cloupickle code path after all, and drilling a bit deeper, I am getting this trace:

```
Traceback (most recent call last):
  File "/Users/charlesstern/Dropbox/pangeo/scratch/distributed/distributed/protocol/pickle.py", line 63, in dumps
    result = pickle.dumps(x, **dump_kwargs)
AttributeError: Can't pickle local object 'ParDo.apply.<locals>.apply_dofn_to_bundle'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/charlesstern/Dropbox/pangeo/scratch/distributed/distributed/protocol/pickle.py", line 68, in dumps
    pickler.dump(x)
AttributeError: Can't pickle local object 'ParDo.apply.<locals>.apply_dofn_to_bundle'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/charlesstern/Dropbox/pangeo/scratch/distributed/distributed/protocol/serialize.py", line 352, in serialize
    header, frames = dumps(x, context=context) if wants_context else dumps(x)
  File "/Users/charlesstern/Dropbox/pangeo/scratch/distributed/distributed/protocol/serialize.py", line 75, in pickle_dumps
    frames[0] = pickle.dumps(
  File "/Users/charlesstern/Dropbox/pangeo/scratch/distributed/distributed/protocol/pickle.py", line 82, in dumps
    result = cloudpickle.dumps(x, **dump_kwargs)
  File "/Users/charlesstern/Dropbox/pangeo/scratch/beam-dask/lib/python3.10/site-packages/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/Users/charlesstern/Dropbox/pangeo/scratch/beam-dask/lib/python3.10/site-packages/cloudpickle/cloudpickle_fast.py", line 632, in dump
    return Pickler.dump(self, obj)
AttributeError: Can't pickle local object '_get_arg_placeholders.<locals>.ArgPlaceholder.__init__'
```
Not sure if I may be missing something basic about how this serialization is expected to work 🤔

**Comment by cisaacstern** [comment]
Continuing think aloud here.. my latest experiments further narrow the cause of the serialization issue to the attempt to pickle `apply_dofn_to_bundle` which is dynamically-defined in the body of `ParDo` in (`runners/dask/transform_evaluator.py`):

- Current state, **raises pickling error** 🥒 ❌ :
    ```diff
      return (
        input_bag
        .map(get_windowed_value, window_fn)
        .map_partitions(apply_dofn_to_bundle)
      )
    ```
- Replace both mapped fn's with identities, **no pickling error** 🥒 ✅ :
    ```diff
      return (
        input_bag
    -   .map(get_windowed_value, window_fn)
    -   .map_partitions(apply_dofn_to_bundle)
    +   .map(lambda x: x)
    +   .map_partitions(lambda x: x)
      )
    ```
- Only replace `get_windowed_value` with identity, **raises pickling error** 🥒 ❌ :
    ```diff
      return (
        input_bag
    -   .map(get_windowed_value, window_fn)
    +   .map(lambda x: x)
        .map_partitions(apply_dofn_to_bundle)
      )
    ```
- Only replace `apply_dofn_to_bundle` with identity, **no pickling error** 🥒 ✅ :
    ```diff
      return (
        input_bag
        .map(get_windowed_value, window_fn)
    -   .map_partitions(apply_dofn_to_bundle)
    +   .map_partitions(lambda x: x)
      )
    ```

Now going to try to lift that definition up to the module level (passing the dynamic aspects as kwargs) and see if that can bring any benefit.

**Comment by cisaacstern** [comment]
> Now going to try to lift that definition up to the module level (passing the dynamic aspects as kwargs) and see if that can bring any benefit.

🥳 😄 This appears to have fixed the serialization issue locally! Cleaning up and will push momentarily!

**Comment by cisaacstern** [comment]
Serialization issues now appear solved, and reverting removal of `client.gather` propagates local errors to the client, therefore giving us the possibility of failing tests, as opposed to to the prior state in which all tests were passing, inclusive of silent errors. So now the "real" development can continue: looks like there are a few failing tests after all, so I'll take a look at fixing those next.

**Comment by cisaacstern** [comment]
The next issue is that GroupByKey is failing in the tests under certain conditions which I suspect have to do with [window merging](https://beam.apache.org/contribute/runner-guide/#window-merging) not working correctly. Digging deeper on that now...

**Comment by cisaacstern** [comment]
> GroupByKey is failing in the tests under certain conditions which I suspect may have to do with [window merging](https://beam.apache.org/contribute/runner-guide/#window-merging) not working correctly.

Confirming that forcing a single partition resolves all of the GBK-related failures in the tests, i.e.

```diff
class Create(DaskBagOp):
  """The beginning of a Beam pipeline; the input must be `None`."""
  def apply(self, input_bag: OpInput) -> db.Bag:
    assert input_bag is None, 'Create expects no input!'
    original_transform = t.cast(_Create, self.transform)
    items = original_transform.values
-   return db.from_sequence(items)
+   return db.from_sequence(items, npartitions=1)
```
I am still wrapping my mind around the relationship between DaskRunner _windows_ and dask bag _partitions_... but I think the latter is the inner implementation of the former? If so, then I think it's fair to say that the remaining work here is, as guessed above, to implement window (i.e., partition) merging here. Continuing on that now...

**Comment by cisaacstern** [comment]
Some quick observations from latest work. I'm seeing at least two issues related to GBK on the DaskRunner. The latter is definitely present on the `2.51.0` release (not just this PR); I'll need to double-check if that's true of the former as well. Eventually some of this (partially the latter issue, which does appear to be a bug in `2.51.0`) can become standalone issue(s):

1. Not all elements of a PCollection of size > 1 are correctly materialized for comparison in the `assert_that` context. This manifests as "missing elements" errors, when in fact the elements that are missing _do_ exist in the PCollection, it's just that they are not being materialized correctly for comparison. This can affect any PCollection of size > 1 (including those for which the user has not invoked GBK), because GBK is used internally in `assert_that`. It is also apparent in the more obvious case when GBK has been invoked in the user's pipeline. In either of these cases, a simple `beam.Map(print)` of the PCollection _does_ reveal all expected elements, but all elements are apparently not available to the `matcher` function in `assert_that` when the comparison is made there. In the `assert_that` case, the code path is:
https://github.com/apache/beam/blob/07d5d534aa8e0f0066e5182a3f17fd9b8a29287c/sdks/python/apache_beam/testing/util.py#L298 https://github.com/apache/beam/blob/07d5d534aa8e0f0066e5182a3f17fd9b8a29287c/sdks/python/apache_beam/transforms/util.py#L258
2. In attempts to debug the above, I was using the [seasonal produce example](https://beam.apache.org/documentation/transforms/python/aggregation/groupbykey/) from the Beam docs, which lead me to discover that the DaskRunner seems to be unable to run GBK using string keys... but other keys such as ints work fine! For example:

    ```python
    # str_keys_gbk.py
  
    import apache_beam as beam
    from apache_beam.options.pipeline_options import PipelineOptions
    from apache_beam.runners.dask.dask_runner import DaskRunner
    from distributed import Client
    
    collections = dict(
        str_keys=[("a", 0), ("a", 1), ("b", 2), ("b", 3)],
        int_keys=[(0, 0), (0, 1), (1, 2), (1, 3)],
    )
    
    if __name__ == "__main__":
        with Client() as c:
            dask_opts = PipelineOptions([f"--dask_client_address={c.scheduler.address}"])
            for runner, options in zip(["DirectRunner", DaskRunner()], [None, dask_opts]):
                for k, v in collections.items():
                    print(f"Running {k} collection on {runner}")
                    with beam.Pipeline(runner=runner) as p:
                        p | beam.Create(v) | beam.GroupByKey() | beam.Map(print)
    
    ```
    ```console
    ➜ python -W ignore str_keys_gbk.py
    Running str_keys collection on DirectRunner
    ('a', [0, 1])
    ('b', [2, 3])
    Running int_keys collection on DirectRunner
    (0, [0, 1])
    (1, [2, 3])
    Running str_keys collection on <apache_beam.runners.dask.dask_runner.DaskRunner object at 0x12dd241f0>
    ('a', [1])
    ('b', [3])
    ('b', [2])
    ('a', [0])
    Running int_keys collection on <apache_beam.runners.dask.dask_runner.DaskRunner object at 0x12dd241f0>
    (1, [2, 3])
    (0, [0, 1])
    ```
    (The warning I'm ignoring here is just a Dask cluster teardown issue, not related.)

**Comment by cisaacstern** [comment]
Progress here is currently blocked by https://github.com/apache/beam/issues/29365. I will shift my focus there now, and return to this once that is resolved.

**Comment by riteshghorse** [comment]
waiting on author

**Comment by cisaacstern** [comment]
I've just merged changes from #29802 (which is a fix for the blocking issue https://github.com/apache/beam/issues/29365 here), and all the tests here are now passing for me locally 🎉 . As noted in #29802, we are currently waiting for https://github.com/dask/dask/pull/10734 to be discussed/reviewed and (hopefully) merged. Once that happens (or a comparable fix is accepted and released in `dask`), next steps are:

- [ ] Update `dask` and `distributed` versions in setup.py over in #29802
- [ ] Merge #29802 (this needs to go in first)
- [ ] Final lint/review for this PR

So to summarize, we are currently paused waiting on upstream review in `dask`. I will re-ping this thread once that discussion progresses. Happy new year, all!

**Comment by alxmrs** [comment]
Wohoo! 🎉 Great work Charles!

**Comment by cisaacstern** [comment]
> and all the tests here are now passing for me locally 🎉

Today I realized that the side input tests that existed in this PR were not actually testing side inputs (they were testing arg/kwarg passing, which IIUC, is not the same thing). So https://github.com/apache/beam/pull/27618/commits/3c0fbc68d2734945b5a560557251dd2c290fc325 makes the side input tests actually test side inputs.

And these tests do **_not_** yet pass. 😄 

(To be clear the upstream `dask` work mentioned in the previous comment is still _totally essential_, insofar as it unlocks the ability to do _any real testing at all_. So the discovery of these test failures is actually a  ✨ *positive development* ✨ in a way, because it reflects progress here: we can finally actually test, and therefore find out what works and what doesn't!)

A fair amount of debugging around led to the skeleton `SideInputMap` draft in https://github.com/apache/beam/pull/27618/commits/8dc89718568ab917ca658ff8db4affcc1de828c1, which is not yet functional but hopefully at least a basis for further consideration.

On a separate note, during debugging I found myself changing the runner as follows, to allow for synchronous execution:

```python
dask_options = options.view_as(DaskOptions).get_all_options(
        drop_default=True)
# comment-out the client so we don't start a 
# thread/process pool during synchronous testing
# client = ddist.Client(**dask_options)

pipeline.replace_all(dask_overrides())

dask_visitor = self.to_dask_bag_visitor()
pipeline.visit(dask_visitor)
opt_graph = dask.optimize(*list(dask_visitor.bags.values()))
# execute synchronously for easier debugging
opt_graph[-1].compute(scheduler="sync")
# futures = client.compute(opt_graph)
# return DaskRunnerResult(client, futures)
```

Possibly there's a way to make that configuration settable via PipelineOptions, and perhaps also as the default setup for `TestPipeline(runner=DaskRunner)` (though we wouldn't want to miss out on all of the multiprocessing bugs that we catch by using a thread/process pool... curious to look into how this is handled in DirectRunner testing).

**Comment by cisaacstern** [comment]
A final note to self for now, which is simple that I am now seeing side input mapping probably needs to happen at the pipeline visitor level... where we have access to context like:

```python
 /.../beam/sdks/python/apache_beam/runners/dask/dask_runner.py(150)visit_transform()
-> inputs = list(transform_node.inputs)
(Pdb) transform_node.inputs
(<PCollection[main.None] at 0x131e7ac50>,)
(Pdb) transform_node.side_inputs
[<apache_beam.pvalue._UnpickledSideInput object at 0x131e58130>]
(Pdb) transform_node
AppliedPTransform(Map(mult_by), ParDo)
(Pdb) op
ParDo(applied=AppliedPTransform(Map(mult_by), ParDo))
(Pdb) op_class
<class 'apache_beam.runners.dask.transform_evaluator.ParDo'>
(Pdb) 
```
Ok... to be continued soon...

**Comment by cisaacstern** [comment]
Status:

- Side inputs now work, with implementation based on Ray Runner
- Added additional side inputs test based on [Ray Runner tests](https://github.com/ray-project/ray_beam_runner/blob/ecc9dba99dc9cbe51c2bbcb1fd472b288a89d1ba/ray_beam_runner/portability/ray_runner_test.py)
- All side inputs tests pass except for 1, which I will take a look at next.

This finally feels like it's getting close.

**Comment by cisaacstern** [comment]
I have now gotten this as far as it can go without the above-referenced fix in dask going in, so going to shift focus back there, and will re-ping here when ready for review.

**Comment by cisaacstern** [comment]
After a hiatus, I am working on this again now.

Today I will complete the upstream fix https://github.com/dask/dask/pull/10734 in dask.

Once that goes in, I'll complete this. Thanks all for your patience.

**Comment by cisaacstern** [comment]
The necessary upstream fix in Dask was merged! 🎉 

Once we get a new Dask release that includes this change, I will finish this PR!

**Comment by alxmrs** [comment]
Wow, @cisaacstern, you've done a heroic effort in getting this change ready for merge! I was thinking of picking this one back up. Is there anything I should going forward? I think that the upstream bugfix has been released?

**Comment by cisaacstern** [comment]
> Wow, @cisaacstern, you've done a heroic effort in getting this change ready for merge! I was thinking of picking this one back up. Is there anything I should going forward? I think that the upstream bugfix has been released?

😃 👋 @alxmrs thanks!! Yes, the upstream fix was merged, and then I _think_ the only/primary thing left was just to pin against a newer release of dask, and get tests to pass here. I dropped the ball on this after that upstream fix went in, as other responsibilities arose. I'd love to pass this back to you and see our mutual efforts land in Beam! Let me know if/how I can support you! Good luck!
